src/main/java/uk/gov/companieshouse/filinghistory/consumer/delta/InternalFilingHistoryApiMapper.java

```diff
@@ -0,0 +1,82 @@
+        this.descriptionValuesMapper = descriptionValuesMapper;
+        this.linksMapper = linksMapper;
+        this.originalValuesMapper = originalValuesMapper;
+    }
+
+    public InternalFilingHistoryApi mapJsonNodeToInternalFilingHistoryApi(final JsonNode topLevelNode, final TransactionKindResult kindResult, final String deltaAt, final String updatedBy) {
+        final JsonNode dataNode = topLevelNode.get("data");
+        final JsonNode originalValuesNode = topLevelNode.get("original_values");
+        final JsonNode descriptionValuesNode = dataNode.get("description_values");
+
+        final String barcode = getValueFromField(topLevelNode, "_barcode");
+        final String documentId = getValueFromField(topLevelNode, "_document_id");
+
+        final String encodedId = kindResult.encodedId();
+        final String companyNumber = getValueFromField(topLevelNode, "company_number");
+
+        return new InternalFilingHistoryApi()
+                .externalData(new ExternalData()
+                        .transactionId(encodedId)
+                        .barcode(barcode)
+                        .type(getValueFromField(dataNode, "type"))
+                        .date(getValueFromField(dataNode, "date"))
+                        .category(getEnumFromField(dataNode, "category", CategoryEnum::fromValue))
+                        .subcategory(getEnumFromField(dataNode, "subcategory", SubcategoryEnum::fromValue))
+                        .description(getValueFromField(dataNode, "description"))
+                        .descriptionValues(descriptionValuesMapper.map(descriptionValuesNode))
+                        .actionDate(getValueFromField(dataNode, "action_date"))
+                        .paperFiled(isPaperFiled(barcode, documentId))
+                        .links(linksMapper.map(companyNumber, encodedId)))
+                .internalData(new InternalData()
+                        .originalValues(originalValuesMapper.map(originalValuesNode))
+                        .originalDescription(getValueFromField(topLevelNode, "original_description"))
+                        .companyNumber(companyNumber)
+                        .parentEntityId(getValueFromField(topLevelNode, "parent_entity_id"))
+                        .entityId(getValueFromField(topLevelNode, "_entity_id"))
+                        .documentId(documentId)
+                        .deltaAt(deltaAt)
+                        .updatedBy(updatedBy)
+                        .transactionKind(kindResult.kind()));
+    }
+
+    private <T extends Enum<?>> T getEnumFromField(final JsonNode node, final String field, Function<String, T> fromValue) {
+        return Optional.ofNullable(node.get(field))
+                .map(JsonNode::textValue)
+                .map(fromValue)
+                .orElse(null);
+    }
+
+    private boolean isPaperFiled(final String barcode, final String documentId) {
```

Comment:
Could this be extracted into its own mapper? It would allow for simpler testing of all the branches (perhaps a parameterised test). I also think when barcode and documentId are null then it should return true which isn't currently happening.